integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -386,3 +409,205 @@ def setUp(self):
+      # TODO(meadowlark): Move this into tf_util.compile_tf_module to prevent
+      # overwritting `reproducer.mlir`.
+      # Disable crash reproducer (to avoid inadvertently overwriting this
+      # path if there are multiple TestCases in the same file).
+      compiler.Context.default_crash_reproducer_path = None
+
+  def setUp(self):
+    # Ran before each unit test.
+    super().setUp()
+    self._ref_module.create_reinitialized()
+    self._tar_modules = [
+        module.create_reinitialized() for module in self._tar_modules
+    ]
+
+  def compare_backends(self, trace_function, rtol=1e-6, atol=1e-6):
+    """Run the reference and target backends on trace_function and compare them.
+
+    Args:
+      trace_function: a function accepting a TracedModule as its argument.
+      rtol: a float as defined in np.allclose. Will be overridden in any
+        comparison between calls to the trace with a `rtol` kwarg.
+      atol: a float as defined in np.allclose. Will be overridden in any
+        comparison between calls to the trace with a `atol` kwarg.
+    """
+    # Trace the test function for each backend.
+    ref_trace = tf_utils.TracedModule(self._ref_module, trace_function)
+    tar_traces = [
+        tf_utils.TracedModule(module, trace_function)
+        for module in self._tar_modules
+    ]
+
+    # Compare each target trace with the reference trace.
+    failed_backend_indices = []
+    for i, tar_trace in enumerate(tar_traces):
+      logging.info("Comparing the reference backend '%s' with '%s'",
+                   ref_trace.backend, tar_trace.backend)
+      traces_match = self._compare_traces(ref_trace, tar_trace, rtol, atol)
+      if not traces_match:
+        failed_backend_indices.append(i)
+
+    # Save the results to disk before validating.
+    trace_dir = os.path.join(self._artifacts_dir, "traces")
+    if not os.path.exists(trace_dir):
+      os.makedirs(trace_dir)
+    ref_trace.save_plaintext(trace_dir, FLAGS.summarize)
+    for tar_trace in tar_traces:
+      tar_trace.save_plaintext(trace_dir, FLAGS.summarize)
+
+    # Validate results.
+    if len(failed_backend_indices) > 0:
+      # Extract info for logging.
+      failed_traces = [tar_traces[b] for b in failed_backend_indices]
+      failed_backends = [trace.backend for trace in failed_traces]
+      failure_info = (
+          "Comparision between the reference backend and the following targets "
+          f"failed: {failed_backends}. The  errors above show the outputs of "
+          "the non-matching calls.")
+
+      # This condition is always True, but is useful for context in the logs.
+      self.assertFalse(len(failed_backends) > 0, failure_info)
+
+  def _compare_traces(self, ref_trace, tar_trace, default_rtol, default_atol):
+    traces_match = True
+    for ref_call, tar_call in zip(ref_trace, tar_trace):
+      ref_call.record_tolerances_if_unset(default_rtol, default_atol)
+      tar_call.record_tolerances_if_unset(default_rtol, default_atol)
```

Comment:
mutating the ModuleCall in place like this seems a bit sketchy. I think we should calls "record_tolerances_if_unset" in the ModuleCall constructor.

Reply:
I think it makes sense for the default tolerance to be defined in the `ModuleCall` via `ModuleCall(method_name, inputs, outputs, rtol=1e-6, atol=1e-6)` if we put the code for comparing two traces into the `TracedModule` class like mentioned above.

We would lose the ability to set the default for an entire trace function via `compare_backends(trace_function, rtol=1e-6, atol=1e-6)`, but the utility of doing that over just specifying the tolerances in each invocation seems low anyway.

Reply:
I think if we move `TracedModule` into `tf_test_utils.py`, then it would make sense for the default test comparison tolerance to be defined in `ModuleCall` via `ModuleCall(method_name, inputs, outputs, rtol=1e-6, atol=1e-6)` instead of `compare_backends()`. This will prevent having to mutate it's state after creation. We'll lose the ability to specify a default tolerance for an entire trace function, but the utility of that over just specifying the tolerance in each invocation seems low anyway.

Reply:
Yeah, I view the tolerance as something that rarely needs to be overridden, and when it does, it's due to a numerical property of the model itself (so there should be no need to change it unless the model is changed).